proposals/NNNN-DXIL-Scalarization.md

```diff
@@ -0,0 +1,90 @@
+`-scalarize-variable-insert-extract` flags  will handle the `BitCastInst`, 
+`CallInst`, `LoadInst` and `StoreInst`, and `GetElementPtrInst` cases. The 
+`AddrSpaceCastInst` case isn't relevant for upstream LLVM.
+
+`MemIntrinsic` is the odd case out. We don't currently emit memcpy for
+the same cases we do in DXC.
+In many cases operations like memcpy and memset seem to happen in global scope.
+See Example:
+- https://hlsl.godbolt.org/z/8vx1Yee7b
+
+In my observation these cases get converted into cbuffer. Since we don't have
+cbuffer support yet we can likely hold off on this cases.
+
+
+The team debated when how early or late this pass should run and three ways to 
+onboard this pass. There was also a discussion on  the scalarizer passes only 
+working on llvm intrinsics and how to allow for DirectX Intrinsics.
+
+### Proposal: When to run pass
+The team determined it was best for this pass to run in the DirectX backend.
+That allows The DirectX backend to be agnostic to the frontend.
+The team also determined the pass should run as late as possible. This has two
+benefits. First code size. if scalarization happens too early things like the 
+`-combiner-alias-analysis` pass limits are reached. More optimizations will be
+possible as analysis passes will benefit from less IR. Second, scalarization 
+for DXIL should be considered a legalization step. To that end it should happen
+at or right before `DXILOpLowering`.
+
+`DXILOpLowering` is also the last place for a functional reason. The scalarizer
+pass only operates on llvm intrinsics that are `TriviallyVectorizable`. Further
+it only converts the scalarized llvm intrinsics meaning there is no way for it
+to know abot DXIL OPs.
+
+### Proposal: Ways to add the Scalarizer Pass
+There are a few ways we could go about this:
+1. support `PassInfoMixin` in the DirectX backend. 
+   - Need to override "buildCodeGenPipeline" for the DirectX backend.
+   - Pro: implementation is easy enough: https://github.com/llvm/llvm-project/compare/main...farzonl:llvm-project:create-codegen-pass-builder
+   - Con: The new PM isn't hooked up in clang yet via `BackendUtil.cpp`'s
+     `AddEmitPasses`. the `TM->addPassesToEmitFile` is all legacy PM.
+2. Add the Legacy Pass manager abstraction for `scalarizer` back
+   - We would be undoing the work the community did to port this pass.
+   - Pro: works seemlessly: https://github.com/llvm/llvm-project/compare/main...farzonl:llvm-project:add-legacy-scalarizer-for-dxil
+3. Fork the `scalarizer` so we can have a legacy and `PassInfoMixin` in the 
+   DirectX Backend?
+   - We could make this work for dxil ops and not just intrinsics.
+   - Con: Mantaining passes like this defeats some of the benefits of the
+     modernization effort.
+
+The team decided the best path forward was to add the legacy pass manager back.
+
+### Proposal: Add Support for DirectX intrinsics
+`isTriviallyScalariable(Intrinsic::ID ID)` will be renamed to `isTriviallyScalarizable`.
```

Comment:
> `isTriviallyScalariable(Intrinsic::ID ID)` will be renamed to `isTriviallyScalarizable`.

Is this just us fixing a typo? Or is "scalariable" a real word and renaming it in some way is changing the meaning?

Reply:
I checked a few spell checkers `Scalariable` isn't a word. `Scalarizable` isn't a word either, but it matches closer to the naming of the pass. My feeling is they mixed `Scalarize` with `Scalable` and dropped the `z`.

Reply:
So the goal of renaming it is just to clarify what it does, not actually any kind of functional change?

Reply:
Correct.